### DISCUSSION TITLE
How to output the singular value of KSP？

### DISCUSSION THREAD

**Discussion by Moyiqi** [idaholab/moose#23936] [Q&A Getting Started] [answered]
When I use -ksp_view_singularvalues , When I use the -ksp_view_singularvalues directive, only gmres can output singularvalues,
other krylov subspace methods like bcgs output 1,How can I get the true singular value？
Do you have any ideas?
Lily

**Comment by GiudGiud** [comment]
Hello

To get singular values we usually use the SVD preconditioner. 

```
[Executioner]
  ...
  petsc_options = '-pc_svd_monitor'
  petsc_options_iname = '-pc_type'
  petsc_options_value = 'svd'
  ...
[]
```

This is very slow, so it's reasonable on 1k dofs, and you could wait a while but still see the result for 10k dofs

Guillaume

**Reply by Moyiqi** [reply]
Thanks,I will try it!
Lily

**Comment by lindsayad** [comment, marked as answer]
[This documentation page](https://petsc.org/release/manualpages/KSP/KSPMonitorSingularValue/#kspmonitorsingularvalue) outlines the ksp types that work with `-ksp_monitor_singular_value`, CG and GMRES. Note that even for these KSP types, at least for GMRES, the estimation will be approximate and improve with the number of linear iterations you take. The options @GiudGiud gave are the best for sufficiently small problems (I usually make sure my dofs are less than ~1000)

**Reply by Moyiqi** [reply]
Thanks,I will try it!
Lily
